feat(stepfunction-tasks): allow ecsruntask to receive dynamically generated task definitions

Michae1CC opened the pull request:
### Issue # (if applicable)

Closes #35307.

### Reason for this change

AWS Step Functions provide users with the ability to synchronously run ECS tasks via the `.sync` integration pattern, see: <https://docs.aws.amazon.com/step-functions/latest/dg/connect-ecs.html>. These changes provide CDK users with the ability to provide the task definition ARN as a JSONata/JSONPath in `aws_stepfunctions_tasks.EcsRunTask` construct with the `.sync` integration pattern. This current isn't possible since the `aws_stepfunctions_tasks.EcsRunTask` construct currently requires the task definition to be provided at the time of deploy via `aws_ecs.TaskDefinition`. This enhancement will benefit users building dynamic ECS workflows in Step Functions.

### Description of changes

Note: These changes allow for complete backwards compatibility and do not effect current usages of  `aws_stepfunction_tasks.EcsRunTaskOptions`.

- A new optional (allows backwards compatibility) parameter `taskDefinitionInput` with type `step_function.TaskInput` has been added to the `aws_stepfunction_tasks.EcsRunTaskOptions` interface. This will provide users with the ability to provide a ECS task definition using a JSONPath/JSONata expression. A validation error is raise if both or neither the `taskDefinition` and `taskDefinitionInput` are provided.
- The `_renderTask` task has been modified to use the `value` property of `taskDefinitionInput` (if the `taskDefinitionInput` is used) when generating the `TaskDefinition`. Otherwise the current implementation of generating the `TaskDefinition` is used when the `taskDefinition` parameter is provided (to maintain backwards compatibility).
- The `taskDefinitionInput` has been typed with `step_function.TaskInput` to remain consistent with how other `aws_stepfunction_tasks` constructs handle parameters that may be JSONPath/JSONata expression (example: <https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_stepfunctions_tasks.SnsPublish.html#message>).
- The parameter name `taskDefinitionInput` has been chosen over  `taskDefinitionArn` since there is a linter rule which does not allow parameters to be prefixed with "Arn".
- A validation error is raise if the user attempts to set both `taskDefinitionInput` and `revisionNumber` since it is impossible to format in a `revisionNumber` if a JSONPath/JSONata expression is provided for the `taskDefinitionInput`.
- New optional (allows backwards compatibility) parameters `networkMode`, `taskRole` and `taskExecutionRole` are used to configure AWS_VPC networking configuration and IAM permissions respectively when `taskDefinitionInput` is used. A validation error is thrown if any of these parameters are set when providing a `taskDefinition` since these parameters should instead be set on the `taskDefinition`, helping reduce implementation logic/conflating.
- The `networkMode` is added since the network mode is currently gleamed from the `taskDefinition`, this cannot be done with `taskDefinitionInput`.
- Currently, the `taskRole` and `taskExecutionRole` values set on the `taskDefinition` are used to create IAM PassRole permissions. The optional  `taskRole` and `taskExecutionRole` provided through the  `aws_stepfunction_tasks.EcsRunTaskOptions` interface are used to reach parity with this particular feature of the `aws_stepfunctions_tasks.EcsRunTask` construct.
- No validation errors are thrown when the `taskRole` and/or `taskExecutionRole` is not provided when using the `taskDefinitionInput` since, strictly speaking, an ECS task requires neither a task role nor a task execution role, see: <https://docs.aws.amazon.com/AmazonECS/latest/developerguide/security-ecs-iam-role-overview.html>

<!--
What code changes did you make? 
Why do these changes address the issue?
What alternatives did you consider and reject?
What design decisions have you made?
-->

### Describe any new or updated permissions being added

- If the `taskRole` and/or `taskExecutionRole` parameters are provided when using the `taskDefinitionInput` then the following statement is added to the step function role policy
```text
{
	actions: ['iam:PassRole'],
	resources: ["<TaskRole and/or TaskExecutionRole if provided>"],
	conditions: {
	  StringEquals: {
		'iam:PassedToService': 'ecs-tasks.amazonaws.com',
	  },
	},
}
```
- Despite the fact that the `'iam:PassedToService'` condition is not present on the `'iam:PassRole'` statement generated when using a `taskDefinition`, this is a huge gain in terms of security since it prevents the step function from potentially passing highly privileged permission to arbitrary principals (see: <https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_iam-condition-keys.html#ck_PassedToService>). The only reason why I have not included the condition changes for the `taskDefinition` case is because these changes would be breaking and is probably best completed as its own change (if preferable to add the changes here let me know and I'll create a new ticket for it).
- When using the `taskDefinitionInput` the following statement is added to the step function role policy
```text
{
  actions: ['ecs:RunTask'],
  resources: ['*'],
  conditions: {
	ArnLike: {
	  'ecs:cluster': this.props.cluster.clusterArn,
	},
  },
}
```
- The resource is starred here since the task definition cannot be determined ahead of time if a JSONPath/JSONata expression is provided.
- Despite the fact that the `'ecs:cluster'` condition is not present on the `'ecs:RunTask'` statement generated when using a `taskDefinition`, I think this is also an improvement to security, preventing the step function from running tasks from any ECS cluster (see: <https://docs.aws.amazon.com/AmazonECS/latest/developerguide/security_iam_id-based-policy-examples.html#IAM_run_policies>). Similarly, the only reason I have not included the condition changes for the `taskDefinition` case is because these changes would be breaking and is probably best completed as its own change. 

<!-- What new or updated IAM permissions are needed to support the changes being introduced? -->

### Description of how you validated changes

- I have added my unit tests to validate new features.
- I have confirmed that the existing integration tests pass without changes.
- I have added `integ.fargate-run-task-taskdef-input`  the integration test. I have confirmed the step function deployed in this integration test executes and completes successfully.

<!-- Have you added any unit tests and/or integration tests? Did you test by hand? -->

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation reviewed (changes requested):
The pull request linter fails with the following errors:

    ❌ Features must contain a change to a README file.
    ❌ The title scope of the pull request should omit 'aws-' from the name of modified packages. Use 'stepfunction-tasks' instead of 'aws-stepfunction-tasks'.

If you believe this pull request should receive an exemption, please comment and provide a justification. A comment requesting an exemption should contain the text `Exemption Request`. Additionally, if clarification is needed, add `Clarification Request` to a comment.

kumvprat commented on packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/ecs/integ.fargate-run-task-taskdef-input.ts:
```diff
@@ -0,0 +1,76 @@
+});
+
+new IntegTest(app, 'SfnTasksEcsFargateRunTaskTest', {
+  testCases: [stack],
```
Can we have some assertions here that check that the step function executes as expected ? 
For reference follow this : https://github.com/Kasra-G/aws-cdk/blob/main/packages/%40aws-cdk-testing/framework-integ/test/aws-stepfunctions/test/integ.distributed-map-redrive.ts#L73-L75, this checks for the execution of the task and also the end result, which here would the execution of the fargate task ?

kumvprat commented on packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/ecs/integ.fargate-run-task-taskdef-input.ts:
```diff
@@ -0,0 +1,76 @@
+});
+const stack = new cdk.Stack(app, 'aws-sfn-tasks-ecs-fargate-run-tasktask-def');
+stack.node.setContext(EC2_RESTRICT_DEFAULT_SECURITY_GROUP, false);
+stack.node.setContext(STEPFUNCTIONS_TASKS_FIX_RUN_ECS_TASK_POLICY, false);
```
The flags below have a default/recommended value `true`, we should test the setup with the default/recommended value : 
- [EC2_RESTRICT_DEFAULT_SECURITY_GROUP](https://github.com/Kasra-G/aws-cdk/blob/main/packages/aws-cdk-lib/cx-api/lib/features.ts#L839-L857) : false
- [STEPFUNCTIONS_TASKS_FIX_RUN_ECS_TASK_POLICY](https://github.com/Kasra-G/aws-cdk/blob/main/packages/aws-cdk-lib/cx-api/lib/features.ts#L1380-L1391) : false

Also I think this not needed as the default value would be false for this : 
`@aws-cdk/aws-lambda:createNewPoliciesWithAddToRolePolicy': false,`

kumvprat replied:
Is idea to allow the new EcsTask changes to work with the default/recommended values ? Or maybe the this feature overall only works in the case that these flags are turned off?

Tell me if I am missing something here(could be that is is a very specific use-case this is trying to solve)

Michae1CC replied:
I initially just copied these from the other tests within the stack, but yes, my specific tests don't really require me to change these values at all. I've removed these lines to revert these flags back to their default values.

Michae1CC commented:
> Thanks for raising the PR @Michae1CC , added a few inline comments

Thanks for the review, I've added in your suggestions.

Michae1CC commented:
> The "not" in both statements is confusing me a bit. The PR does add these specific conditions as `Condition` in the step-function role policy. Can you provide more details around the reasoning ?

I mean to say that the _current_ implementation does not include IAM condition checks to the `iam:PassRole` and `ecs:RunTask` statements (generated for step function execution role to run the tasks when providing a `taskDefinition` to the construct). My changes do include these condition checks but only when a `taskDefinitionInput` is provided.

The condition check I've added to the `iam:PassRole` statements restricts the step function to only passing the role to the ECS service. Without this check, the step function may pass the ECS execution role and/or ECS task execution role to arbitrary principles.

The condition check I've added to the `ecs:RunTask` statements restricts the step function to only run tasks for the provided ECS cluster. Without this check, the step function may run ECS tasks from an arbitrary ECS cluster.

> Also can you include the proper document that captures this ECS and step-function integration. That would be helpful as reference for the changes being done. Is this the [correct link ? ](https://docs.aws.amazon.com/step-functions/latest/dg/connect-ecs.html)

Yes that is the correct link.

Michae1CC commented:
> Can you also add the link in the PR description so that it's easy to understand the context ? (the documentation link that documents details about ECS and step-function integration)

I've added the link

Michae1CC commented on packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:
I did this so that these parameters couldn't be omitted. I noticed in design guidelines that union types may not be supported for CDK targeted languages so I've changed them to be optional.

Michae1CC commented on packages/aws-cdk-lib/aws-stepfunctions-tasks/test/ecs/run-tasks.test.ts:
No, I've added checks IAM Policy checks

Michae1CC commented on packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:
> Are the union types needed ?

As above, changed to be optional.

> Also in the validation message can we have a check that prints something like NetworkMode is not defined when the value of the variable is undefined.

Done, I've updated the unit tests.

kumvprat commented on packages/@aws-cdk-testing/framework-integ/test/aws-stepfunctions-tasks/test/ecs/integ.fargate-run-task-taskdef-input.ts:
```diff
@@ -0,0 +1,93 @@
+  cpu: 256,
+});
+
+taskDefinition.addContainer('TheContainer', {
```
@Michae1CC I have a functionality question

The idea behind having the JSONata/JSONPath resolve to the task definition arn is that the task definition can be created dynamically as a prequisite to running the task via the step function.
 
To emulate your use-case : The task definition could be created as step function step and then passed on to next step which would run it.

The current test gives an impression that the task definition will be available to deployment-time which is a subset of all the cases this PR is trying to cover, since this PR is also covering the case where the task definition will not be available till runtime i.e when the sfn StateMachine runs.

kumvprat replied:
I am open to discussing this as the step functions can be used in variety of scenarios like your custom scenario mentioned on the [issue](https://github.com/aws/aws-cdk/issues/35307#issue-3345071420). This seems to be a truly dynamic use-case

Michae1CC replied:
I had originally hard coded the task definition arn input since really the only functionality we are testing here are that changes to the `aws_stepfunctions_tasks.EcsRunTask` construct should allow users to provide the task definition arn as a string (the construct doesn't really care if it was dynamically generated or not).

That said, having the task definition arn perhaps better demonstrates the intended/a more practical use of this feature. I've changed the integration test to have the task definition dynamically generated through a `RegisterTaskDefinition` action.

kumvprat commented on packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:
I see a lot of : `@default $`

Can you update with the default values, maybe : `@default - None` or add a bit more explanation in the comments around what happens when the parameter value is not passed?

kumvprat replied:
> I think something like 'EcsRunTaskFromTaskDefArnOptions' would be more suitable.
I think `TaskDefFromArnOptions` might be a better fit.

In any case whatever name we choose, my suggestions are below : 


Using below : 
**OPTION 1**
> That way we can have simpler validations that check that exactly one object is present between taskDefinition and dynamicTaskDefinition :
>
>interface TaskDefFromArnOptions {
>/**
>   * The family and revision (family:revision) or full ARN of the task
>   * definition to run. Cannot be used with `taskDefinition`.
>   *
>   * @default $
>   */
>  readonly taskDefinitionInput: sfn.TaskInput;
>/**
>   * The ECS task networking mode.
>   *
>   * @default $
>   */
>  readonly networkMode?: ecs.NetworkMode;
>
 > /**
 >  * The ECS task role.
 >  *
 >  * @default $
>   */
>  readonly taskRole?: iam.IRole;
>
>  /**
>   * The ECS task execution role.
>   *
>   * @default $
>   */
>  readonly taskExecutionRole?: iam.IRole;
>}

and then use it in `EcsRunTaskOptions` : 

> interface EcsRunTaskOptions {
> ....
> readonly taskDefinition?: ecs.TaskDefinition;
> ....
> readonly taskDefFromArnOptions?: sfn.TaskDefFromArnOptions;
> .....
> }

**OR**
**OPTION 2**
rename the properties that are supposed to work together with some similar prefix (like `taskDefFromArnNetworkMode` where `taskDefFromArn` is the prefix to be used for other properties)


**For the default value comments, please follow : https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md#defaults**

Michae1CC replied:
Okay that makes, I think I might run with option 2 (even though it differs from the implementation described in the issue) since it means we're not doing as many checks and it provides the affordance of these parameters needing to be set together.

Also, I appreciate the prompt reviews!

Michae1CC commented on packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:
> But I don't find any documentation regarding this, did you come across any documentation for this ?

I'm not sure exactly what documentation you're referring. This policy statement is required for the step function to run ECS tasks with the prescribed task execution role and task role.

- Documentation on setting up a stepfunction IAM permissions to run ECS tasks: <https://docs.aws.amazon.com/step-functions/latest/dg/connect-ecs.html#ecs-iam>

> Also can we abstract out these into two functions, one for static and one for dynamic task definition creation ? It is harder to read/comprehend what's happening in different cases here.

I've refactored the code for policy statement generation.

kumvprat reviewed (commented):
Let an inline comment : https://github.com/aws/aws-cdk/pull/35347#discussion_r2398052129

Michae1CC commented:
I've squashed everything into a single commit since GH seems to be struggling with the diffs.

Michae1CC commented on packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:
For classes that implement `IEcsLaunchTarget`, the `taskDefinition` parameter is only used to check that the task definition is compatible with the `launchTarget`. I've made the `taskDefinition` parameter optional for when a `taskDefinitionFromArnOptions` is provided. Since we can't query `taskDefinitionFromArnOptions` for `launchTarget` compatibility, if a `taskDefinition` is not provided (i.e. `taskDefinitionFromArnOptions`), I simply omit the compatibility check.

I think the tests are still accurate since, when this construct is used with `taskDefinitionFromArnOptions` it just skips this `launchTarget` compatibility check.

kumvprat replied:
Ah okay I had a look again the `undefined` check allows it to run without taskDefinition being provided.
Thanks for explanation

So when we work with `arn` we have no way of validating before deployment that the task definition that is going to be used is compatible with the `LaunchTargetOptions` being provided ? 

I think that can be the fair trade-off since this use-case is supporting already existing `taskDefinition` or one created dynamically where in both cases it can be assumed that the user knows about `taskDefinition` compatibility with the launch target types.

Michae1CC commented:
I've gone with this approach since adding a new parameter to the construct interface is a common method for adding new functionality, for example the `aws_cdk.aws_lambda.Function` has introduced `applicationLogLevelV2` parameter to provide a better way to set logging configuration. Is using an construct interface to provide information for dynamically built resources a common pattern/used anywhere else? If not, could we get a second opinion?

Michae1CC commented:
This will involve a lot more work for this specific feature. If there aren't really any existing examples of this in CDK, can we get a second opinion on modifying the ITaskDefinition interface to be used in this way?

aws-cdk-automation commented:
<!--STALE PR - CHANGES REQUESTED-->
This PR has been in the CHANGES REQUESTED state for 3 weeks, and looks abandoned. Note that PRs with failing linting check or builds are not reviewed, please ensure your build is passing 

To prevent automatic closure:
- Resume work on the PR
- OR request an exemption by adding a comment containing 'Exemption Request' with justification e.x "Exemption Request: <justification>"
- OR request clarification by adding a comment containing 'Clarification Request' with a question e.x "Clarification Request: <question>"

This PR will automatically close in 14 days if no action is taken.
